### PR TITLE
fix: reduce size of docker image by leveraging a multi-stage build

### LINKED ISSUE #175: Docker images are large
The official ksops docker images are quite large.
They seem to contain the entire `GOPATH` including full source code for all dependencies (now totaling 1.8GB!).

I'd suggest creating a multi-stage dockerfile with minimal things present in the final image.

### PR DESCRIPTION
**Pull request by therealdwright**
The current docker image builds to about 2.7GB on my machine, this
can be improved through leveraging multi-stage builds in docker.

With this commit, the new image size will be less than 200mb and
the necessary binaries are copied to /usr/local/bin

Closes #175

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -24,12 +25,19 @@ RUN export GOOS=$(echo ${TARGETPLATFORM} | cut -d / -f1) && \
 
 WORKDIR /go/src/github.com/viaduct-ai/kustomize-sops
 
-ADD . .
+COPY go.mod go.sum ./
+RUN go mod download
 
-# Perform the build
+COPY . .
```
Comment (devstein) [inline Dockerfile:31]: What's the purpose of `COPY`ing the go.mod go.sum over twice?

Wouldn't this suffice

```Dockerfile
COPY . .
RUN go mod download
# rest
```

**Review by devstein — COMMENTED** [review]
Thank you for making a PR! One quick question

File: Dockerfile (hunk 1 above)
Comment (therealdwright) [reply Dockerfile:31]: I was initially thinking we could have the dependencies in an early layer and copy the code after, but I think it's probably reasonable to just copy everything and do a go mod download. I'll push a fix now.

**Review by devstein — APPROVED** [review]
LGTM

File: Dockerfile [hunk 2]
```diff
@@ -4,7 +4,8 @@ ARG GO_VERSION="1.19"
 #--------Build KSOPS and Kustomize-----------#
 #--------------------------------------------#
 
-FROM golang:$GO_VERSION
+# Stage 1: Build KSOPS and Kustomize
+FROM golang:$GO_VERSION AS builder
 
 LABEL org.opencontainers.image.source="https://github.com/viaduct-ai/kustomize-sops"
```
Comment (james-callahan) [inline Dockerfile:10]: The `LABEL` needs to be moved down to the final image

Comment (therealdwright) [reply Dockerfile:10]: https://github.com/viaduct-ai/kustomize-sops/pull/181

**Comment by Piroddi** [comment]
Hi, this PR has made the Custom ArgoCD Image doc outdated:

https://github.com/viaduct-ai/kustomize-sops#custom-argo-cd-w-ksops-dockerfile

**Comment by therealdwright** [comment]
> Hi, this PR has made the Custom ArgoCD Image doc outdated:

Thanks for flagging @Piroddi - I'll push a PR to fix the docs today.

**Pull request merged**